### PR TITLE
gui: add xattr filter editor (fixes #8660)

### LINKED ISSUE #8660: GUI editor for xattr filter patterns
The per-folder xattr filter is a complex object which is not by itself editable in the advanced config editor, and there's nothing custom built for it in the normal settings editor, so it's currently unavailable except for hacking the config file manually.

There should be an editor for it in the normal settings dialog, under the xattr enablement checkboxes. There is should be possible to add a list of patterns, each with an allow/deny checkbox.

The default is to allow all attributes if there are no patterns configured; this should be indicated in the GUI. When there are any patterns at all, the default (for things not matched by any pattern) becomes deny; this should also be indicated, especially if the user just adds one deny-pattern -- we should then suggest adding a `permit *` pattern at the end, maybe even add it by default and allow it to be removed or edited, or something like that.

### PR DESCRIPTION
**Pull request by er-pa**
### Purpose

Added a GUI element in the advanced folder configuration panel to be able to manipulate the extended attribute filter settings. 

I intended to keep it as simple as possible, no automatic additions or other magic behind the screens. Mostly because I expect that the users who will use this will know what they are doing. But also because I'm not a fan myself of automatically added rules/lines when manipulating settings, unnecessary complexity. However, some information will still be shown when appropriate;
- Default behaviour if a wild-card/any is not present as last element
- A reminder in case that only deny-rules exist while the default in that case is deny - suggesting the addition of a permit-any rule in the end.

Otherwise empty rules are being cleaned up on saving the configuration and new entries are being pushed before a wild-card/any rule, if that one is present as last element.

Any further logic/validation is not present, on purpose.

### Testing

- Manipulate a folder's xattr filter using the new part in the GUI -> check result in the config file.
- Manipulate the default folder configuration -> add a new folder -> verify that the settings are correct for the newly added folder (and default settings).

### Screenshots

Screenshot 2022-12-28 at 11 51 46
Screenshot 2022-12-28 at 11 51 54
Screenshot 2022-12-28 at 11 52 03

### Documentation

Supporting documentation is prepared as well, https://github.com/syncthing/docs/pull/779

### CODE REVIEW

**Comment by tomasz1986** [comment]
I've had a quick look just at the GUI side of things.

1. Would it be possible not to use arrow functions?  They're not used anywhere else in the GUI, and using them here breaks the whole Web GUI in older browsers that don't support them.
2. How about moving the whole "Checking the check-box..." `help-block` up between the "Xattr Filter" and "Active filter rules" labels? Alternatively, right below "Active filter rules"? Doing so should match the convention used in the rest of the GUI and also eliminate the need for the `hr`.
3. How about only showing this section when syncing/sending extended attributes have actually been enabled (checked)?

File: gui/default/syncthing/folder/editFolderModalView.html [hunk 1]
```diff
@@ -323,7 +323,64 @@
               </p>
             </div>
           </div>
+
+          <div class="row">
+            <div class="col-md-12">
+              <p>
+                <label translate>Xattr Filter</label>
```
Comment (tomasz1986) [inline gui/default/syncthing/folder/editFolderModalView.html:330]:
```suggestion
                <label translate>Extended Attributes Filter</label>
```

The abbreviation isn't used anywhere else in the GUI, so I think it's better to keep things in "human" language whenever possible 😉. The advanced Docs and the code are another story, of course.

**Comment by er-pa** [comment]
> I've had a quick look just at the GUI side of things.
> 
> 1. Would it be possible not to use arrow functions?  They're not used anywhere else in the GUI, and using them here breaks the whole Web GUI in older browsers that don't support them.
> 2. How about moving the whole "Checking the check-box..." `help-block` up between the "Xattr Filter" and "Active filter rules" labels? Alternatively, right below "Active filter rules"? Doing so should match the convention used in the rest of the GUI and also eliminate the need for the `hr`.
> 3. How about only showing this section when syncing/sending extended attributes has actually been enabled (checked)? Just to remove the unnecessary clutter from the GUI for those that don't use these features.

1.) Probably. But the reasoning I'm not entirely behind. If I'm not wrong any browser post-2016 has support for this, we talk about almost 7 years without updates. How far back do we go?...can't use this argument till eternity ;)
2.) I've tried a lot in that area, also with a drop-down box instead of checkbox to make that entire help-box obsolete. So that's surely an option and something to consider.
3.) Uhm, yeah, that's absolutely a possibility

**Comment by tomasz1986** [comment]
> 1.) Probably. But the reasoning I'm not entirely behind. If I'm not wrong any browser post-2016 has support for this, we talk about almost 7 years without updates. How far back do we go?...can't use this argument till eternity ;)

For me, it's mainly about Internet Explorer. The current Web GUI supports IE11 which used to be the only browser bundled into some editions of Windows (e.g. Windows Server and the whole Enterprise LTSB/C line). This state has only started to change last and this year (with Enterprise LTSC 2021 and Server 2022 which come with Edge, although IE11 is still present). I don't think it's worth to break the perfectly functioning GUI in those systems just yet. Just for the record, IE11 still receives security updates, although it shouldn't really matter that much in this case, as we're talking about controlling a local application here.

There's also the Android app but that's probably not important anymore, as support for pre-4.4 Android has recently been dropped in it, and the Android 5+ WebView does seem to support arrow functions.

**Comment by er-pa** [comment]
Textual update:

- Arrow-functions have been converted
- Explanatory text moved upwards
- Xattr renamed
- Attempted to fix the 'out of bound' of the delete-button. Can't reproduce it here on any device or screen-size, so I'm not sure if it's fixed.

File: gui/default/vendor/bootstrap/css/bootstrap.css [hunk 2]
```diff
@@ -1649,7 +1649,7 @@ input[type="search"] {
 input[type="radio"],
 input[type="checkbox"] {
   margin: 4px 0 0;
-  margin-top: 1px \9;
```
Comment (tomasz1986) [inline gui/default/vendor/bootstrap/css/bootstrap.css:1652]: Why touch this one? The whole `\9` thing is an old hack for IE 7/8/9 (see https://stackoverflow.com/questions/8004765/what-is-the-meaning-of-9-in-this-width-property-value-declaration). It should be irrelevant for other browsers.

File: gui/default/syncthing/core/syncthingController.js [hunk 3]
```diff
@@ -3330,6 +3331,84 @@ angular.module('syncthing.core')
...
+            // When the user explicitely added a wild-card, we don't show hints.
+            if (filterEntries.length === 1 && filterEntries[0].match === '*') {
+                return '';
+            }
+            // If all the filter entries are 'deny', we suggest adding a permit-any
+            // rule in the end since the default is already deny in that case.
+            if (filterEntries.every(function (entry) {
+                return entry.permit === false;
+            })) {
+                return  $translate.instant('Hint: only deny-rules detected while the default is deny. Consider adding `permit any` as last rule.');
```
Comment (tomasz1986) [inline gui/default/syncthing/core/syncthingController.js:3381]:
```suggestion
                return  $translate.instant('Hint: only deny-rules detected while the default is deny. Consider adding "permit any" as last rule.');
```

I think it's better to stick to `"`. Backticks aren't really used like that outside of Markdown, and the GUI currently uses standard quotes for code already, e.g.

image

Ideally, this should be in `<code></code>`, but HTML tags don't go well inside strings that are supposed to be translated.

**Comment by tomasz1986** [comment]
> Textual update:
> 
> * Arrow-functions have been converted
> * Explanatory text moved upwards
> * Xattr renamed
> * Attempted to fix the 'out of bound' of the delete-button. Can't reproduce it here on any device or screen-size, so I'm not sure if it's fixed.

Thank you for all the fixes! Everything GUI-wise seems to work as expected (including in IE11) with the exception for the last item.

image

Not a deal breaker, as the buttons are still functional. Just for the record, I test in Chromium using DevTools and toggling the device toolbar to switch to phone/tablet mode. Then, I set the screen size either to iPhone SE or responsive and shrink it horizontally. I would test and try to find the culprit to this one myself but I'll only have the time around the weekend or so.

**Review by calmh — APPROVED** [review]

**Pull request merged**